### PR TITLE
[RLlib] Introduce SpecDict data structure 

### PR DESCRIPTION
**Pull request by kouroshHakha**
## Why are these changes needed?
This is needed for defining io specs on the models.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/models/specs/specs_dict.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                elements of the spec, but can contain more entries.
+        Raises:
+            ValueError: If the data doesn't match the spec.
+        """
+        data = NestedDict(data)
+        missing_keys = set(self.keys()).difference(set(data.keys()))
+        if missing_keys:
+            raise ValueError(_MISSING_KEYS_FROM_DATA.format(missing_keys))
+        if exact_match:
+            data_spec_missing_keys = set(data.keys()).difference(set(self.keys()))
```
Comment (gjoliver) [inline rllib/models/specs/specs_dict.py:48]: buffer the keys in some local variable?
then kind of feel like missing_keys and data_spec_missing_keys are not necessary.

File: rllib/models/specs/specs_dict.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            if data_spec_missing_keys:
+                raise ValueError(_MISSING_KEYS_FROM_SPEC.format(data_spec_missing_keys))
+
+        for spec_name, spec in self.items():
+            data_to_validate = data[spec_name]
+            if isinstance(spec, TensorSpecs):
+                spec.validate(data_to_validate)
+            elif isinstance(spec, Type):
```
Comment (gjoliver) [inline rllib/models/specs/specs_dict.py:56]: we need a doc string for this spec type.
how do you construct this? and what's the intended use?

File: rllib/models/specs/specs_dict.py (hunk 1 above)
Comment (kouroshHakha) [reply rllib/models/specs/specs_dict.py:48]: even if I cache the keys to improve performance I still need A-B vs. B-A right?

Comment (gjoliver) [reply rllib/models/specs/specs_dict.py]: yes. A-B and B-A are fine.
I missed that you need the diffs for error messages.
please ignore, just cache the keys.

File: rllib/models/specs/specs_dict.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/models/specs/specs_dict.py:120]: updated the doctring. Thanks for the reminder.

**Review by gjoliver — APPROVED** [review]

**Comment by gjoliver** [comment]
Merging. Test failures should be related.

**Pull request merged**